design/0005-circuit-breaking.md

```diff
@@ -0,0 +1,231 @@
+
+---
+
+Protocol Specific features -
+
+1. For HTTP/gRPC based communication, users should be able to configure circuit
+   breaking either per endpoint or whole service.
+   [Example](https://drive.google.com/open?id=1JK65u9Mnmkmy_X8fczeh6RO6O-vNiTtY)
+   User can configure ejection based on many different metrics. These metrics
+   would be calculated over a given time range.
+   - Success Rate
+   - Percentile Latencies
+   - Max Requests
+   - Consecutive Failures
+2. For TCP based communication -
+   [Example](https://drive.google.com/open?id=1yAAurygsZQ-r58TEA0QiFf9Sz15jqRJV)
+
+---
+
+[0] The term 'traffic' means any kind of communication happening between two
+services. It can be HTTP requests or gRPC messages over HTTP or L4 TCP
+connections. But the circuit breaking functionality itself won't be protocol
+agnostic. Its implementation would be protocol specific and it's upto community
+which protocol they want to add support for and how circuit breaking will make
+sense in that protocol.
+
+## Design proposal
+
+[design-proposal]: #design-proposal
+
+Right now if I just consider ServiceProfile spec, IMO it seems to represent more
+of profile-for-http-service rather than profile-for-generic-service. May be it
+makes sense given that it supports only HTTP and grpc over HTTP as of now. Given
+current architecture, I feel it's best to provide circuit breaking functionality
+through service profiles. That means it would be limited to above two kind of
+services as of now. Circuit breaking logic would be applied in the data plane
+based on the metrics recorded inside the proxy.
+
+Since circuit breaking needs to be handled per protocol, we will leave TCP for
+now and can revisit the same in future.
+
+In future, for any other protocol support for circuit breaking, one will first
+need to incorporate the protocol into ServiceProfile itself and then implement
+the circuit breaking functionality.
+
+For HTTP/gRPC, we already have golden metrics available in the proxy. Along with
+that, we can add additional metrics and use those for circuit breaking logic. In
+case no instance is available to serve request due to circuit breaking, then we
+will return 503 for HTTP with some custom header and gRPC message for gRPC.
+
+ServiceProfile interface would have following additions for the HTTP and gRPC -
+
+### Circuit Breaker
+
+```yaml
+routes:
+  - name: "/foo"
+    circuitBreaker:
+      # Circuit Breaker Kind
+      # (Required)
+      kind: ...
+
+      # Max percentage of instances that can be ejected.
+      # Default - 100%
+      maxEjectionPercentage: 10
+
+      # The min duration of time in seconds for which an instance is marked
+      # as ejected.
+      # (Required)
+      baseEjectionTime: 60
+
+      # Ejection Time Inflation
+      # How to increase ejection time on every consecutive ejection.
+      # (Optional. Default - None)
+      ejectionTimeInflation: ...
```

Comment:
> Why would someone need to configure this? What use case does it support?

Same as above. Though I understand that having lot of switches can be confusing for the end user but we can add them up with sane defaults so that it's win-win situation for both non-advanced and fine tuners.

Reply:
But I'm fine with keeping it simple as well and add it in future if need be.